lld/ELF/InputSection.cpp

```diff
@@ -918,7 +917,8 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
       continue;
 
     if (tombstone ||
-        (isDebug && (type == target.symbolicRel || expr == R_DTPREL))) {
+        ((isDebug && (type == target.symbolicRel || expr == R_DTPREL))) ||
```

Comment:
Would be nice to know from @MaskRay (or anyone else) why this `type == target.symbolicRel` is here, and if there's some way we can generalize it to cover the `.debug_names` case we're interested in (a `R_X86_64_32` relocation) without needing to special case `isDebugNames` in this condition

Reply:
Yes. :D As I mentioned, for DWARF64 isDebugNames check is not necessary.

Reply:
While we're waiting for an expert opinion, if youre bored, you could try removing the `type == target.symbolicRel` check and running the lld tests, about 5 tests fail - investigating one fo those failures to see what goes wrong in the absence of that condition might help point the way to understanding how to generalize the condition enough to cover the `.debug_names` situation, without generalizing too far and breaking those tests

Reply:
Well basically that condition checks if relocation is 64bit: R_X86_64_64 (type == 1).
Primarily used in .debug_addr. Incidentally why DWARF64 tests works without adding isDebugNames to that condition. since for .debug_names relocation type is controlled by DWARF32 vs DWARF64.
Slightly more interesting case is expr == R_DTPREL. Since pretty much all relocs in debug sections are ABS. Looks like that relocation is used to reference TLS symbols. Learned something new today.

Reply:
Yeah, mechanically I understand what the check does. 

What I mean is: Why is this check necessary? What bad things happen (some tests fail, but I haven't looked closely at what they're testing or what the bad behavior looks like) when this check fails/isn't checked? And why is it OK to ignore this check in the `.debug_names` case? And could we look at that reason and generalize the check so it's not specifically about `.debug_names`, but about whatever general property is true in the `.debug_names` case and would be true in other cases?

Basically, it seems the `type == target.symbolicRel` check is overly restrictive (we have at least one counter example - the `.debug_names` `R_X86_64_32` situation) and I'd like to know ideally how to generically relax this check to account for our situation, but also others that may come up in the future/just not to have a weird `.debug_names` special case that we don't really understand/can't explain why it's acceptable here, but not there - and layering special cases makes the code harder to understand and maintain.

Reply:
Ah, I see what you mean. Yeah was going to dig in further today and see how to generalize. You are right it is over restrictive. Ideally it probably should be something like if (tombstone || debug), and then some logic depending on relocation type and probably section. I am hesitant to change current defaults due to downstream tools (cough lldb), breaking or something.